File: cli/rust/blackjack/src/dealer.rs
/*
	Blackjack

	Copyright 2015 Sam Saint-Pettersen
	Released under the MIT/X11 License
*/

use cards::Cards;

extern crate rand;
use self::rand::Rng;

/// Blackjack Dealer.
pub struct Dealer {
	cards: Vec<String>,
    values: Vec<u16>,
}

impl Dealer {

	pub fn new() -> Dealer {
		Dealer {
			cards: Vec::new(),
            values: Vec::new(),
		}
	}

    pub fn calc_total(&self) -> u16 {
        let mut values = self.values.clone();
        values.sort();
        values.reverse();
        let mut total: u16 = 0;
        for i in values {
            let mut v: u16 = 0;
            if i == 1 {
                if total + 11 <= 21 {
                    v = 11
                }
                else if total + 11 > 21 {
                    v = 1
                }
            }
            total += v;
        }
        total
    }

    pub fn has_blackjack(&self) -> bool {
        let mut blackjack = false;
        if self.calc_total() == 21 && self.cards.len() == 2{
            println!("Dealer has Blackjack!");
            blackjack = true
        }
        blackjack
    }

    pub fn is_bust(&self) -> bool {
        let mut bust = false;
        if self.calc_total() > 21 {
            println!("Dealer is bust!");
            bust = false
        }
        bust
    }

    fn process_cards(&mut self, cards: Vec<&str>) {
        let v = cards[1].parse::<u16>().ok();
        let value = match v {
            Some(v) => v,
            None => 0
        };
        self.cards.push(cards[0].to_string());
        self.values.push(value);
    }

    pub fn receive_cards(&mut self, player_cards: Vec<String>) {
        for card in player_cards {
            let cvs = card.split(":");
            let cv: Vec<&str> = cvs.collect();
            self.process_cards(cv);
        }
    }

    /*fn hit(&mut self, &mut cards: Cards) {
        let card = cards.draw();
        self.cards.push("foo".to_string());
        println!("Dealer hits.");
        println!("Dealer gets {}", card);
        self.values.push(cards.get_value());
    }*/

    fn stand(&self) {
        println!("Dealer stands.");
    }

    pub fn shuffle(&mut self, cards: Cards) {
        if cards.get_played() == 0 || cards.get_played() >= 45 {
            println!("-----------------------------------------------");
            println!("Dealer is shuffling cards...");
            println!("-----------------------------------------------");
            cards.shuffle();
        }
    }

    pub fn deal(&mut self, cards: Cards) -> Vec<String> {
        let dealt = Vec::new();
        let i = 0;
        println!("-----------------------------------------------");
        println!("Dealer is dealing cards for a new game...");
        println!("-----------------------------------------------");
        while i < 2 * 2 {
            let d = format!("{}:{}", cards.draw(), cards.get_value());
            dealt.push(d.to_string());
            i += 1
        }
        i = 0;
        while i < 2 {
            let cvs = dealt[i].split(":");
            let cv: Vec<&str> = cvs.collect();
            let v = cv[1].parse::<u16>().ok();
            let value = match v {
                Some(value) => value,
                None => 0
            };
            self.cards.push(cv[0].to_string());
            self.values.push(value);
            i += 1
        }
        println!("Dealer has:");
        println!("[**]{}", self.cards[1]);
        dealt
    }

    /*pub fn respond(&mut self, cards: Cards) {
        self.show_cards();
        loop {
            let mut total = 0;
            while total <= 18 {
                total = self.calc_total();
                if total == 16 {
                    if rand::thread_rng().gen_range(0, 5) >= 3 {
                        //self.hit(cards); // Take risk.
                    }
                    else {
                        self.stand(); // Play it safe.
                        break;
                    }  
                }
                else if total >= 17 {
                    self.stand();
                    break;
                }
                else {
                    //self.hit(cards);
                }
            }
        }
    }*/

    pub fn show_cards(&mut self) -> u16 {
        let mut cards = String::new();
        let vcards = self.cards.clone();
        for card in vcards {
            cards = format!("{}{}", cards, card);
        }
        println!("Dealer has:");
        println!("{} --> {}", cards, self.calc_total());
        self.calc_total()
    }
}

File: cli/go/blackjack/dealer.go
/*
	Blackjack

	Copyright 2016 Sam Saint-Pettersen
	Released under the MIT/X11 License.
*/

package main

import (
	"fmt"
	"math/rand"
	"time"
	"strings"
	"strconv"
	"sort"
)

type Dealer struct {
	cards []string
	values []int
}

func newDealer() Dealer {
	return Dealer {
		cards: make([]string, 0),
		values: make([]int, 0),
	}
}

func (d Dealer) _CalcTotal() int {
	sort.Sort(sort.Reverse(sort.IntSlice(d.values)))
	total := 0
	for _, v := range d.values {
		if v == 1 {
			if total + 11 <= 21 {
				v = 11
			} else {
				v = 1
			}
		}
		total += v
	}
	return total
}

func (d Dealer) _HasBlackjack() bool {
	blackjack := false
	if d._CalcTotal() == 21 {
		fmt.Println("\nPlayer has Blackjack!")
		blackjack = true
	}
	return blackjack
}

func (d Dealer) _IsBust() bool {
	bust := false
	if d._CalcTotal() > 21 {
		fmt.Println("\nPlayer is bust!")
		bust = true
	}
	return bust
}

func (d Dealer) _Hit(cards Cards) (Cards, Dealer) {
	cards, card := cards.Draw()
	d.cards = append(d.cards, card)
	fmt.Println("Dealer hits.")
	fmt.Println("Dealer gets", card)
	cards, value := cards.GetValue()
	d.values = append(d.values, value)
	fmt.Println("Player has", d._CalcTotal())
	return cards, d
}

func (d Dealer) _Stand() {
	fmt.Println("Player stands.")
	fmt.Println("Player has", d._CalcTotal())
}

func (d Dealer) Shuffle(cards Cards) Cards {
	if cards.GetPlayed() == 0 || cards.GetPlayed() >= 45 {
		fmt.Println("-------------------------------------------------------")
		fmt.Println("Dealer is shuffling cards...")
		fmt.Println("-------------------------------------------------------")
		cards.Shuffle()
	}
	return cards
}

func (d Dealer) Deal(cards Cards) (Cards, []string) {
	dealt := make([]string, 0)
	fmt.Println("-------------------------------------------------------")
	fmt.Println("Dealer is dealing cards for a new game...")
	fmt.Println("-------------------------------------------------------")
	for i := 0; i < 2 * 2; i++ {
		cards, card := cards.Draw()
		cards, value := cards.GetValue()
		dealt = append(dealt, fmt.Sprintf("%s%s", card, value))
	}
	for i := 0; i < 2; i++ {
		cv := strings.Split(dealt[i], ":")
		d.cards = append(d.cards, cv[0])
		value, _ := strconv.Atoi(cv[1])
		d.values = append(d.values, value)
	}
	fmt.Println("\nDealer has:")
	fmt.Println(fmt.Sprintf("[**]%s", d.cards[1]))
	return cards, dealt[2:]
}

func (d Dealer) Respond(cards Cards) (Cards, Dealer) {
	d.ShowCards()
	for {
		total := 0
		for total <= 18 {
			total = d._CalcTotal()
			if total == 16 {
				rand.Seed(int64(time.Now().Nanosecond()))
				if rand.Intn(6) >= 3 {
					cards, d = d._Hit(cards)
				} else {
					d._Stand()
					break
				}
			} else if total >= 17 {
				d._Stand()
				break
			} else {
				cards, d = d._Hit(cards)
			}
		}
	}
	return cards, d
}

func (d Dealer) ShowCards() int {
	cards := ""
	for _, card := range d.cards {
		cards = fmt.Sprintf("%s%s", cards, card)
	}
	fmt.Println("\nDealer has:")
	fmt.Println(fmt.Sprintf("%s --> %d", cards, d._CalcTotal()))
	return d._CalcTotal()
}
